### PR TITLE
[FLINK-22121][table-planner-blink] FlinkLogicalRankRuleBase now check if name of rankNumberType already exists in the input

### PR DESCRIPTION
**Pull request by TsReaper**
## What is the purpose of the change

Currently names of rank fields are all `w0$o0`, so if the input of a `Rank` is another `Rank` some exception will occur. To solve this, we should check if name of rank field has occurred in the input in `FlinkLogicalRankRuleBase`.

## Brief change log

- FlinkLogicalRankRuleBase now check if name of rankNumberType already exists in the input

## Verifying this change

This change added tests and can be verified by running the added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/FlinkLogicalRankRule.scala [hunk 1]
```diff
@@ -98,6 +98,12 @@ abstract class FlinkLogicalRankRuleBase
       case _ => throw new TableException(s"Unsupported rank function: $rankFun")
     }
 
+    // this is to avoid duplicated column name with the input
+    var rankNumberName = rankNumberType.get.getName
+    while (window.getInput.getRowType.getFieldNames.contains(rankNumberName)) {
+      rankNumberName += "$"
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/FlinkLogicalRankRule.scala:104]: When hitting duplicate field names, Calcite ususally append `_<id>` after it, e.g. `$g_0`, `$g_1`. 
What do you think about `rankNumberName += "_0"`?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchPhysicalOverAggregateRule.scala [hunk 2]
```diff
@@ -190,7 +190,8 @@ class BatchPhysicalOverAggregateRule
     val inputNameList = inputType.getFieldNames
     val inputTypeList = inputType.getFieldList.asScala.map(field => field.getType)
 
-    val aggNames = aggCalls.map(_.getName)
+    // we should avoid duplicated names with input column names
+    val aggNames = aggCalls.map(c => RowTypeUtils.getUniqueName(c.getName, inputNameList))
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchPhysicalOverAggregateRule.scala:194]: what if there are three duplicate names: "a, a, a", the result will be "a, a0, a0". 
the new unique name should be added to inputNameList.

please add an UT for this class

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchPhysicalOverAggregateRule.scala [hunk 3]
```diff
@@ -26,14 +26,14 @@ import org.apache.flink.table.planner.plan.nodes.logical.FlinkLogicalOverAggrega
...
 import org.apache.flink.table.planner.plan.utils.{AggregateUtil, OverAggregateUtil, SortUtil}
-
 import org.apache.calcite.plan.RelOptRule._
 import org.apache.calcite.plan.{RelOptCluster, RelOptRule, RelOptRuleCall}
 import org.apache.calcite.rel._
 import org.apache.calcite.rel.`type`.RelDataType
 import org.apache.calcite.rel.core.Window.Group
 import org.apache.calcite.rel.core.{AggregateCall, Window}
 import org.apache.calcite.tools.ValidationException
+import org.apache.flink.table.planner.typeutils.RowTypeUtils
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchPhysicalOverAggregateRule.scala:36]: nit: reorder the imports

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchPhysicalOverAggregateRule.scala (hunk 2 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchPhysicalOverAggregateRule.scala:194]: please add a plan test for this change

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution @TsReaper , I left some comments

**Comment by TsReaper** [comment]
@flinkbot run azure

**Review by godfreyhe — APPROVED** [review]

**Pull request closed without merging**